### DISCUSSION TITLE
How to extract gradients from a tensor and add it to the data of another tensor?

### DISCUSSION THREAD

**Discussion by shettysach** [huggingface/candle#2343] [Q&A]
Taken from Karpathy's lecture - [makemore_part1_bigrams.ipynb](https://github.com/karpathy/nn-zero-to-hero/blob/master/lectures/makemore/makemore_part1_bigrams.ipynb)

```python
  # backward pass
  W.grad = None # set to zero the gradient
  loss.backward()
  
  # update
  W.data += -50 * W.grad
```

Here W and loss are PyTorch tensors. 
I'm having trouble understanding GradStore. How can I implement the same using candle tensors?

**Comment by shettysach** [comment]
```rust
    let mut weights = Var::randn(0_f32, 1_f32, &[27, 27], &DEVICE)?;

    for k in 0..50 {
        let loss = forward_pass(&xs, &ys, &weights)?;
        println!("Iteration {k} - loss = {}", loss.to_scalar::<f32>()?);

        weights.backward()?.remove(&weights);
        let loss_grad = loss.backward()?;
        let weights_grad = loss_grad.get(&weights).unwrap();

        weights = Var::from_tensor(
            &weights
                .broadcast_sub(&weights_grad.broadcast_mul(&Tensor::new(&[50_f32], &DEVICE)?)?)?,
        )?;
    }
```

Currently using this, but am unsure if this is the best way.
